Always show security screen and shows error page when trying to access forbidden data-source

DarshitChanpura opened the pull request:
### Description
When a user with multi-data-source feature enabled tries to access the cluster, the Security tab will be rendered based on whether the user has access to connected OpenSearch cluster, and not on whether they have access to remote cluster. This is an inconsistent behavior. To fix this we will display Security tab always when `dataSource.enabled` feature flag is set to true.

Changes addressed:
1. Shows security screen when MDS is enabled.
2. When a user doesn't have access to said backend cluster, a "Permission error" component will be displayed.
3. Ensures uniform loading experience across page loads and data-source switches.
4. Handles malformed dataSource object in the url

### Category
New feature

### Why these changes are required?
To promote seamless experience for multi-data-source users.


### What is the old behavior before changes and new behavior after changes?
Before, admins of remote data-source could not access the security configuration features via dashboard.
With this change, admins will have config screen access to clusters it has access to. If they don't have access, a message along the lines `You do not have permission to view the data for <cluster-name>.` will be displayed.


### Issues Resolved
- Resolves https://github.com/opensearch-project/security-dashboards-plugin/issues/1962

### Testing
- Manual + Automated

**1. Only access to Remote DataSource using admin credentials**
- MDS Disabled

https://github.com/opensearch-project/security-dashboards-plugin/assets/35282393/dcb4af89-4980-4973-b122-2d16f2502a8e

- MDS Enabled
Note: All refreshes that you seen in this video are manual refreshes to show that correct data-source is set in the url.

https://github.com/opensearch-project/security-dashboards-plugin/assets/35282393/2003cabc-e9dd-4731-8f37-6893a9fc5bea


**2. Only access to Local DataSource (also remote data-source using admin credentials)**
- MDS Disabled

https://github.com/opensearch-project/security-dashboards-plugin/assets/35282393/491fc4df-1380-48ad-9b1c-03d4599a2921

- MDS Enabled

https://github.com/opensearch-project/security-dashboards-plugin/assets/35282393/ba2d5f18-1d91-4201-aec7-125e73676764

__Currently, the remote data-source connection uses stored credentials to interact with remote data-source.__

### Check List
- [x] New functionality includes testing
~- [ ] New functionality has been documented~
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

derek-ho commented on public/plugin.ts:
```diff
@@ -96,7 +97,7 @@ export class SecurityPlugin
     );
 
-    if (apiPermission) {
+    if (mdsEnabled || apiPermission) {
```
Nice! 🐐

derek-ho commented on public/apps/configuration/panels/user-list.tsx:
```diff
@@ -213,67 +214,71 @@ export function UserList(props: AppDependencies) {
-        {deleteConfirmModal}
-      </EuiPageContent>
+      {errorFlag ? (
+        <AccessErrorComponent dataSourceLabel={dataSource && dataSource.label} />
```
I am not sure if we can always assume that the error flag means MDS issue, it could also be network issue/whatever was the list of issues possible before MDS was enabled WDYT?

DarshitChanpura replied:
Addressed.

derek-ho commented on public/apps/configuration/panels/tenant-list/tenant-list.tsx:
```diff
@@ -45,9 +45,8 @@ export function TenantList(props: TenantListProps) {
-        await setIsMultiTenancyEnabled(
-          (await getDashboardsInfo(props.coreStart.http)).multitenancy_enabled
-        );
+        const dashboardsInfo = await getDashboardsInfo(props.coreStart.http);
+        setIsMultiTenancyEnabled(dashboardsInfo?.multitenancy_enabled || false);
```
lint? Or why was this introduced?

DarshitChanpura replied:
for better readability. The original code was awaiting on updating the state, because it internally waited on dashboardsinfo request to complete. I have separated the two out

derek-ho reviewed (commented):
Had a general question - can you also fix the tests?

cwperks commented on public/apps/access-error-component.tsx:
```diff
@@ -0,0 +1,32 @@
+  message?: string;
+}
+
+export const AccessErrorComponent: React.FC<AccessErrorComponentProps> = (props) => {
```
Will this page also be rendered when MDS is disabled and a user tries to directly navigate to a security URL without having the proper access?

i.e. if a user goes directly to `http://localhost:5601/app/security-dashboards-plugin#/users` by typing it into the URL instead of navigating with the menu.

When I try to directly go to a security url as a non-admin I currently get Application Not Found.

DarshitChanpura replied:
we do not register security plugin if either MDS is not enabled or user doesn't have sufficient permissions. Hence, Application not found will still be displayed in this case where MDS is disabled.

cwperks commented on public/apps/configuration/panels/role-list.tsx:
```diff
@@ -119,15 +120,14 @@ export function RoleList(props: AppDependencies) {
         setErrorFlag(false);
       } catch (e) {
-        console.log(e);
         setErrorFlag(true);
```
I think this per-page setting of the error flag makes it possible to selectively view pages for users granted `restapi:admin/*` permissions. Currently, the logic to show `Security` on the main menu or not only takes into account users who have been assigned one of the `plugins.security.restapi.roles_enabled` roles.

DarshitChanpura replied:
not sure what you meant by this one. Can you elaborate?

cwperks replied:
i.e. a User with permission to restapi:admin/internaluser should be able to view the list of internal users in dashboards, but get "You do not have permission..." for other screens.

Currently, its all or none. You can either see "Security" in the main menu or you cannot.

cwperks commented on public/utils/datasource-utils.ts:
```diff
@@ -27,7 +27,12 @@ export function getClusterInfo(dataSourceEnabled: boolean, cluster: DataSourceOp
 export function getDataSourceFromUrl(): DataSourceOption {
   const urlParams = new URLSearchParams(window.location.search);
   const dataSourceParam = (urlParams && urlParams.get(DATASOURCEURLKEY)) || '{}';
```
nit: Why no explicitly check if `urlParams.has(DATASOURCEURLKEY)` here and return empty object/null/undefined to indicate to the caller that the request does not contain dataSource?

~~return JSON.parse('{}');~~ `return {};`

DarshitChanpura replied:
the urlParam needs to be parsed to a valid JSON entity. If we directly return without proper parsing, the page will crash, which is what I'm trying to solve with this change.

derek-ho commented on public/apps/configuration/app-router.tsx:
```diff
@@ -300,7 +276,24 @@ export function AppRouter(props: AppDependencies) {
                 }}
               />
-              {getTenancyRoutes()}
+              {multitenancyEnabled && (
```
TY

derek-ho commented on public/apps/configuration/panels/auth-view/auth-view.tsx:
```diff
@@ -40,17 +43,24 @@ export function AuthView(props: AppDependencies) {
+        if (e.response && [400, 403].includes(e.response.status)) {
+          setAccessErrorFlag(true);
+        }
+        setErrorFlag(true);
```
Should this be else? Or doesn't matter too much right?

DarshitChanpura replied:
doesn't matter

DarshitChanpura replied:
it is a generic error flag that should be set regardless of what type of error is caught. I've added an error flag specific to access errors to handle the display of the message.

derek-ho commented on public/apps/configuration/panels/auth-view/auth-view.tsx:
```diff
@@ -76,18 +98,26 @@ export function AuthView(props: AppDependencies) {
-      {/* @ts-ignore */}
-      <AuthorizationPanel authz={authorization} loading={loading} config={props.config} />
+      {loading ? (
+        <EuiLoadingContent />
```
Isn't this case already being handled by the AccessErrorComponent?

DarshitChanpura replied:
for this specific component and how it is loaded, I had to handle it here. Also, these components do not render loading screens so the change between data-sources is not seamless

derek-ho commented on public/apps/configuration/panels/auth-view/auth-view.tsx:
```diff
@@ -76,18 +98,26 @@ export function AuthView(props: AppDependencies) {
+        <AccessErrorComponent loading={loading} dataSourceLabel={dataSource && dataSource.label} />
+      ) : (
+        <>
+          {/* @ts-ignore */}
```
What is being ignored here?

DarshitChanpura replied:
not sure, it was already present.

derek-ho commented on public/apps/configuration/panels/permission-list/permission-list.tsx:
```diff
@@ -210,17 +213,22 @@ export function PermissionList(props: AppDependencies) {
+      if (e.response && [400, 403].includes(e.response.status)) {
+        setAccessErrorFlag(true);
+      }
       setErrorFlag(true);
```
else?

DarshitChanpura replied:
https://github.com/opensearch-project/security-dashboards-plugin/pull/1964#discussion_r1617675526

derek-ho commented on public/apps/configuration/panels/permission-list/permission-list.tsx:
```diff
@@ -361,57 +369,68 @@ export function PermissionList(props: AppDependencies) {
-        </EuiPageBody>
-      </EuiPageContent>
+      {loading ? (
+        <EuiLoadingContent />
```
Handled by component?

DarshitChanpura replied:
it is handled by component only when there is an error

derek-ho commented on public/apps/configuration/panels/tenant-list/configure_tab1.tsx:
```diff
@@ -176,28 +181,38 @@ export function ConfigureTab1(props: AppDependencies) {
+          default_tenant: defaultTenant = '',
+        } = dashboardsInfo;
         await setOriginalConfiguration({
-          multitenancy_enabled: (await getDashboardsInfo(props.coreStart.http))
-            .multitenancy_enabled,
-          private_tenant_enabled: (await getDashboardsInfo(props.coreStart.http))
-            .private_tenant_enabled,
-          default_tenant: (await getDashboardsInfo(props.coreStart.http)).default_tenant,
```
🐐

derek-ho commented on public/apps/configuration/panels/tenant-list/configure_tab1.tsx:
```diff
@@ -306,6 +321,19 @@ export function ConfigureTab1(props: AppDependencies) {
     </EuiText>
   );
+
+  if (loading) {
+    return <EuiLoadingContent />;
+  }
```
Same comment as above

DarshitChanpura replied:
https://github.com/opensearch-project/security-dashboards-plugin/pull/1964#discussion_r1617678368

derek-ho commented on public/apps/configuration/panels/tenant-list/manage_tab.tsx:
```diff
@@ -484,6 +491,17 @@ export function ManageTab(props: AppDependencies) {
     );
   };
 
+  if (loading) {
+    return <EuiLoadingContent />;
+  }
```
Same as above

DarshitChanpura replied:
https://github.com/opensearch-project/security-dashboards-plugin/pull/1964#discussion_r1617678368

derek-ho commented on public/apps/configuration/panels/user-list.tsx:
```diff
@@ -213,67 +221,73 @@ export function UserList(props: AppDependencies) {
-        </EuiPageBody>
-        {deleteConfirmModal}
-      </EuiPageContent>
+      {loading ? (
+        <EuiLoadingContent />
+      ) : accessErrorFlag ? (
```
Same as above

DarshitChanpura replied:
https://github.com/opensearch-project/security-dashboards-plugin/pull/1964#discussion_r1617678368

derek-ho reviewed (approved):
Mostly LGTM, thanks for the thorough testing and videos for the change 👍

derek-ho commented on public/apps/configuration/panels/service-account-list.tsx:
```diff
@@ -111,8 +114,14 @@ export function ServiceAccountList(props: AppDependencies) {
+        // requests with existing credentials but insufficient permissions result in 403, remote data-source requests with non-existing credentials result in 400
+        if (e.response && [400, 403].includes(e.response.status)) {
+          setAccessErrorFlag(true);
+        }
```
Maybe shallow is not waiting for the component to re-render and show your custom component

derek-ho replied:
You may need to use mount instead: https://enzymejs.github.io/enzyme/docs/api/mount.html

derek-ho replied:
Personally I wouldn't block on this codecov, but good to verify the change

DarshitChanpura replied:
> Maybe shallow is not waiting for the component to re-render and show your custom component

i don't think so, since tests would fail if component didn't render correctly. Its something else

DarshitChanpura commented:
@cwperks @derek-ho I'm marking conversations as resolved. Please re-open them if any comment requires more discussion.

scrawfor99 commented on public/apps/configuration/panels/tenant-list/test/__snapshots__/tenant-list.test.tsx.snap:
```diff
@@ -1,5 +1,19 @@
+exports[`Tenant list AccessError component should load access error component: configure tab 1`] = `
+<AccessErrorComponent
+  dataSourceLabel="Local cluster"
+  message="You do not have permissions to configure tenancy"
```
```suggestion
  message="You do not have permission to configure tenancy"
```
Same

DarshitChanpura replied:
https://github.com/opensearch-project/security-dashboards-plugin/pull/1964#discussion_r1621348751

scrawfor99 commented on public/apps/configuration/panels/tenant-list/test/__snapshots__/tenant-list.test.tsx.snap:
```diff
@@ -1,5 +1,19 @@
+exports[`Tenant list AccessError component should load access error component: manage tab 1`] = `
+<AccessErrorComponent
+  dataSourceLabel="Local cluster"
+  message="You do not have permissions to manage tenants"
```
```suggestion
  message="You do not have permission to manage tenants"
```

Same

DarshitChanpura replied:
https://github.com/opensearch-project/security-dashboards-plugin/pull/1964#discussion_r1621348751

scrawfor99 commented on public/apps/configuration/panels/test/__snapshots__/role-list.test.tsx.snap:
```diff
@@ -1,5 +1,169 @@
+            >
+               
+              (
+              0
```
Nit: Is this supposed to be split up like this?

DarshitChanpura replied:
auto generated by jest. so I think yes

scrawfor99 commented on public/apps/configuration/panels/test/__snapshots__/role-list.test.tsx.snap:
```diff
@@ -1,5 +1,169 @@
+          color="subdued"
+          size="xs"
+        >
+          Roles are the core way of controlling access to your cluster. Roles contain any combination of cluster-wide permission, index-specific permissions, document- and field-level security, and tenants. Then you map users to these roles so that users gain those permissions. 
```
```suggestion
          Roles are the core way of controlling access to your cluster. Roles contain any combination of cluster-wide permissions, index-specific permissions, document- and field-level security, and tenants. Roles are then mapped to users in order to provide the associated permissions. 
```

DarshitChanpura replied:
This text is what we display now and should not be altered. If you think this text should be changed, please open an issue.

scrawfor99 reviewed (approved):
Just some minor notes.

DarshitChanpura commented on public/apps/configuration/access-error-component.tsx:
```diff
@@ -0,0 +1,35 @@
+  const {
+    loading = false,
+    dataSourceLabel,
+    message = 'You do not have permissions to view this data',
```
There are generally multiple permissions involved, like `read`, `search`,  `write` etc. Hence I left it as plural

DarshitChanpura commented on public/apps/configuration/panels/auth-view/auth-view.tsx:
```diff
@@ -40,17 +43,24 @@ export function AuthView(props: AppDependencies) {
+  }, [props.coreStart.http, dataSource]);
 
-  if (isEmpty(authentication)) {
+  if (isEmpty(authentication) && !loading) {
```
loading will never be null since the default value during state initialization is set to false

DarshitChanpura commented on public/apps/configuration/panels/auth-view/auth-view.tsx:
```diff
@@ -76,18 +98,26 @@ export function AuthView(props: AppDependencies) {
           <h1>Authentication and authorization</h1>
         </EuiTitle>
-        {props.config.ui.backend_configurable && (
+        {!loading && !errorFlag && props.config.ui.backend_configurable && (
```
not needed as neither of these values be null

DarshitChanpura commented on public/apps/configuration/panels/audit-logging/audit-logging.tsx:
```diff
@@ -93,6 +93,16 @@ function renderStatusPanel(onSwitchChange: () => void, auditLoggingEnabled: bool
+    <AccessErrorComponent
+      loading={loading}
+      dataSourceLabel={dataSource && dataSource.label}
+      message="You do not have permissions to configure audit logging settings"
```
https://github.com/opensearch-project/security-dashboards-plugin/pull/1964#discussion_r1621348751

DarshitChanpura commented on public/apps/configuration/panels/audit-logging/test/__snapshots__/audit-logging.test.tsx.snap:
```diff
@@ -644,3 +646,39 @@ exports[`Audit logs render when AuditLoggingSettings.enabled is true 1`] = `
+  <EuiSpacer />
+  <AccessErrorComponent
+    loading={false}
+    message="You do not have permissions to configure audit logging settings"
```
https://github.com/opensearch-project/security-dashboards-plugin/pull/1964#discussion_r1621348751

DarshitChanpura commented on public/apps/configuration/panels/auth-view/test/__snapshots__/auth-view.test.tsx.snap:
```diff
@@ -0,0 +1,31 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`Auth view should load access error component 1`] = `
```
1st snapshot, it's a syntax that jest uses. These were auto-generated by jest.

DarshitChanpura commented on public/apps/configuration/panels/tenant-list/configure_tab1.tsx:
```diff
@@ -306,6 +321,19 @@ export function ConfigureTab1(props: AppDependencies) {
+    return (
+      <AccessErrorComponent
+        dataSourceLabel={LocalCluster.label}
+        message="You do not have permissions to configure tenancy"
```
https://github.com/opensearch-project/security-dashboards-plugin/pull/1964#discussion_r1621348751

DarshitChanpura commented on public/apps/configuration/panels/tenant-list/manage_tab.tsx:
```diff
@@ -484,6 +491,17 @@ export function ManageTab(props: AppDependencies) {
+    return (
+      <AccessErrorComponent
+        dataSourceLabel={LocalCluster.label}
+        message="You do not have permissions to manage tenants"
```
https://github.com/opensearch-project/security-dashboards-plugin/pull/1964#discussion_r1621348751

DarshitChanpura commented:
@scrawfor99 addressed all comments and resolved conversations that were duplicates. Please resolve or comment on the open items so we can address those and bring them to closure.

cwperks commented:
@DarshitChanpura Small nit on design. Can the "You do not have permissions text" be centered in the container?

cwperks commented:
The change LGTM otherwise. Thank you @DarshitChanpura!

DarshitChanpura commented:
@cwperks The alignment is handled by parent component in this case. EuiPageContent defaults to [center justified](https://github.com/opensearch-project/oui/blob/main/src/components/page/page_content/page_content.tsx#L43-L44). If we want to align it correctly we will have to adjust parent component, since all the wrapper component align the text as left justified.
IMO, it is fine to leave it as is and get the PR merged since making changes to parent component is not feasible.
